File: 1593-split-a-string-into-the-max-number-of-unique-substrings/1593-split-a-string-into-the-max-number-of-unique-substrings.java
class Solution {
    HashSet<String> vis = new HashSet<>();
    
    int fun(int ind, String s)
    {
        if(ind >= s.length()) return 0;
        
        String temp = "";
        int ans = 0;
        for(int k=ind; k<s.length(); k++)
        {
            temp = temp + s.charAt(k);
            if(!vis.contains(temp))
            {
                vis.add(temp);
                ans = Math.max(ans, 1 + fun(k+1, s));
                vis.remove(temp);
            }
        }
        return ans;
    }
    public int maxUniqueSplit(String s) {
        vis.add("");
        return fun(0, s);
    }
}

File: 1593-split-a-string-into-the-max-number-of-unique-substrings/1593-split-a-string-into-the-max-number-of-unique-substrings.py
class Solution:
    def __init__(self):
        self.vis = set()  
    
    def maxUniqueSplit(self, s: str) -> int:
        
        def fun(ind: int, string: str) -> int:
            if ind >= len(string):
                return 0
            
            temp = ""
            ans = 0
            for i in range(ind, len(string)):
                temp += string[i]
                if temp not in self.vis:  
                    self.vis.add(temp)
                    ans = max(ans, 1 + fun(i+1, string))
                    self.vis.remove(temp)  
            return ans
        
        return fun(0, s)  
